docs/src/main/asciidoc/hibernate-search-guide.adoc

```diff
@@ -0,0 +1,627 @@
+
+== Automatic import script
+
+For the purpose of this demonstration, let's import an initial dataset.
+
+Let's create a `src/main/resources/import.sql` file with the following content:
+
+[source,sql]
+----
+INSERT INTO author(id, firstname, lastname) VALUES (nextval('hibernate_sequence'), 'John', 'Irving')
+INSERT INTO author(id, firstname, lastname) VALUES (nextval('hibernate_sequence'), 'Paul', 'Auster')
+
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'The World According to Garp', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'The Hotel New Hampshire', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'The Cider House Rules', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'A Prayer for Owen Meany', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'Last Night in Twisted River', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'In One Person', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'Avenue of Mysteries', 1);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'The New York Trilogy', 2);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'Mr. Vertigo', 2);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'The Brooklyn Follies', 2);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'Invisible', 2);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), 'Sunset Park', 2);
+INSERT INTO book(id, title, author_id) VALUES (nextval('hibernate_sequence'), '4 3 2 1', 2);
+----
+
+== Preparing the infrastructure
+
+We need a PostgreSQL instance and an Elasticsearch cluster.
+
+Let's use Docker to start one of each:
+
+[source, shell]
+----
+docker run -it --rm=true --name elasticsearch_quarkus_test -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:7.0.1
+----
+
+[source, shell]
+----
+docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 --name postgresql_quarkus_test -e POSTGRES_USER=quarkus_test -e POSTGRES_PASSWORD=quarkus_test -e POSTGRES_DB=quarkus_test -p 5432:5432 postgres:11.3
+----
+
+== Time to play with your application
+
+You can now interact with your REST service:
+
+ * start Quarkus with `./mvnw compile quarkus:dev`
+ * open a browser to `http://localhost:8080/`
+ * search for authors or book titles (we initialized some data for you)
+ * create new authors and books and search for them too
+
+As you can see, all your updates are automatically synchronized to the Elasticsearch cluster.
+
+== Building a native executable
+
+You can build a native executable with the usual command `./mvnw package -Pnative`.
```

Comment:
Running this executes tests, which may not work if the containers are not up. See my comment on the other PR.

Reply:
Also, you should probably remind that GraalVM must be set up and link to some documentation.

Reply:
Also it didn't work for me with GraalVM 19. I get this:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:native-image (default) on project hibernate-search-elasticsearch: Failed to generate a native image: Failed to build native image: Cannot run program "/home/yrodiere/tools/java/graalvm19//bin/native-image" (in directory "/home/yrodiere/workspaces/reviews/quarkus-hsearch/target"): error=2, No such file or directory -> [Help 1]
```

And indeed:
```
$ ls /home/yrodiere/tools/java/graalvm19//bin/            
appletviewer  gu    jar        javac    javap         jconsole  jhat   jmap        js         jstat      keytool       node  pack200     rmic         schemagen   tnameserv  wsimport
clhsdb        hsdb  jarsigner  javadoc  java-rmi.cgi  jdb       jinfo  jps         jsadebugd  jstatd     lli           npm   policytool  rmid         serialver   unpack200  xjc
extcheck      idlj  java       javah    jcmd          jdeps     jjs    jrunscript  jstack     jvisualvm  native2ascii  orbd  polyglot    rmiregistry  servertool  wsgen
```

I probably did something wrong, but that's just more proof we need a bit of guidance :)

Reply:
* GraalVM is in the prerequisites at the top of the page. I added a link.
 * I explicitly require starting the Docker containers above and it shouldn't start the tests as there are no tests here
 * GraalVM 19 is not yet supported by Quarkus. They broke a lot of things between rc16 and 19...

Reply:
> it shouldn't start the tests as there are no tests here

Yes there are, created by the archetype: 

```
$ tree src/test/        
src/test/
└── java
    └── org
        └── acme
            └── hibernate
                └── search
                    └── elasticsearch
                        ├── LibraryResourceTest.java
                        └── NativeLibraryResourceIT.java

6 directories, 2 files
```

And I get:

```
[INFO] --- maven-surefire-plugin:2.22.0:test (default-test) @ hibernate-search-elasticsearch ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.hibernate.search.elasticsearch.LibraryResourceTest
2019-05-29 12:57:32,240 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-05-29 12:57:33,514 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 1274ms
2019-05-29 12:57:36,774 INFO  [org.hib.sea.map.orm.mas.mon.imp.SimpleIndexingProgressMonitor] (Hibernate Search: identifierloader - 1) HSEARCH000027: Going to reindex 2 entities
2019-05-29 12:57:36,883 INFO  [org.hib.sea.map.orm.mas.mon.imp.SimpleIndexingProgressMonitor] (Hibernate Search: identifierloader - 1) HSEARCH000027: Going to reindex 13 entities
2019-05-29 12:57:37,248 INFO  [org.hib.sea.map.orm.mas.mon.imp.SimpleIndexingProgressMonitor] (main) HSEARCH000028: Reindexed 15 entities
2019-05-29 12:57:37,251 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 3.536s. Listening on: http://[::]:8081
2019-05-29 12:57:37,251 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, hibernate-search-elasticsearch, jdbc-postgresql, narayana-jta, resteasy, resteasy-jsonb]
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.785 s <<< FAILURE! - in org.acme.hibernate.search.elasticsearch.LibraryResourceTest
[ERROR] testHelloEndpoint  Time elapsed: 0.684 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Expected status code <200> but was <404>.

	at org.acme.hibernate.search.elasticsearch.LibraryResourceTest.testHelloEndpoint(LibraryResourceTest.java:17)

2019-05-29 12:57:37,952 INFO  [io.quarkus] (main) Quarkus stopped in 0.006s
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   LibraryResourceTest.testHelloEndpoint:17 1 expectation failed.
Expected status code <200> but was <404>.

[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

Even if you talk about starting the docker containers before, I personally don't expect a postgresql and elasticsearch container to be necessary to package the application. Maybe I'm just weird...

Reply:
Just to clarify, I get the error above even with the postgresql/elasticsearch containers up and running. It's because we removed the method being tested as part of the guide.

Reply:
Thanks, switching to rc16 did the trick. It works, and it's quite amazing.

Note I had to use `-DskipTests` to avoid the problem above.

I also had close running apps because compilation would fail with exit code 137. It means out of memory, a shame there's no explicit message but I guess it's just the way it is.

More importantly, having the Elasticsearch and PostgreSQL containers up and running while we launch this command might be a problem depending on the user configuration. I have 16GB of RAM and still had issues with the 7.7GB required by GraalVM for compiling. Maybe a comment about requirements (a lot of ram, ...) might help? Or just a link to a section of the quarkus documentation that already exists? I'm sure that kind of thing is already documented somewhere.

Reply:
I added a note.